File: Engine/script/runtimelibrary/Scene.cs
/****************************************************************************
Copyright (c) 2011-2013,WebJet Business Division,CYOU
 
http://www.genesis-3d.com.cn

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System;
using System.Runtime.CompilerServices;

namespace ScriptRuntime
{
    /// <summary>
    /// 场景类
    /// 主要用户查找当前场景Actor等功能
    /// </summary>
    public partial class Scene : Base
    {
        private Scene(DummyClass__ dummyObj)
        {
            // empty
        }
        private Scene()
        {
            
        }
        ~Scene()
        {
            ICall_Scene_Release(this);
        }
        /// <summary>
        /// 获取场景中Actor的数量
        /// </summary>
        public int ActorCount
        {
            get
            {
                return ICall_Scene_GetActorCount(this);
            }
        }
        /// <summary>
        /// 根据Actor列表的索引,返回对应的Actor
        /// </summary>
        /// <param name="index">Actor列表中的索引</param>
        /// <returns>返回对应Actor</returns>
        public Actor GetActor(int index)
        {
            return ICall_Scene_GetActor(this, index);
        }
        /// <summary>
        /// 根据fastID返回场景中其对应的Actor
        /// </summary>
        /// <param name="fastID">fastID的值</param>
        /// <returns>返回对应Actor</returns>
        public Actor FindActor(uint fastID)
        {
            return ICall_Scene_FindActorByFastID(this, fastID);
        }
         /// <summary>
        /// 根据tagID返回场景中其对应的Actor
        /// </summary>
        /// <param name="tagID">tagID的值</param>
        /// <returns>返回对应Actor</returns>
        public Actor FindActorInGroup(uint tagID)
        {
            return ICall_Scene_FindActorByTag(this, tagID);
        }

        /// <summary>
        /// 根据GroupID返回场景中其对应Actor列表
        /// </summary>
        /// <param name="GroupID">GroupID的值</param>
        /// <returns>返回对应的Actor列表</returns>
        public Actor[] FindActorsInGroup(uint tagID)
        {
            return ICall_Scene_FindActorsByTag(this, tagID);
        }
        /// <summary>
        /// 根据guid返回场景中其对应的Actor
        /// </summary>
        /// <param name="guid">guid的值</param>
        /// <returns>返回对应Actor</returns>
        public Actor FindActor(Guid guid)
        {
            byte[] guidarray = guid.ToByteArray();
            return ICall_Scene_FindActorByGuid(this, guidarray);
        }
        /// <summary>
        /// 根据Actor的名字返回对应Actor
        /// </summary>
        /// <param name="name">Actor的名字</param>
        /// <returns>返回对应的Actor</returns>
        public Actor FindActor(String name)
        {
            return ICall_Scene_FindActorByName(this, name);
        }
        /// <summary>
        /// 给场景添加一个Actor
        /// </summary>
        /// <param name="obj">要添加的Actor实例</param>
        /// <returns>添加成功返回true,反之返回false</returns>
        public bool AddActor(Actor obj)
        {
            return ICall_Scene_AddActor(this, obj);
        }
        /// <summary>
        /// 移除场景中对应的Actor实例
        /// </summary>
        /// <param name="obj">要移除的Actor实例</param>
        /// <returns>移除成功返回true,反之返回false</returns>
        public bool RemoveActor(Actor obj)
        {
            return ICall_Scene_RemoveActor(this, obj);
        }

    }
}

File: Engine/app/appframework/scene.cc
/****************************************************************************
Copyright (c) 2011-2013,WebJet Business Division,CYOU
 
http://www.genesis-3d.com.cn

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
#include "stdneb.h"
#include "appframework/scene.h"
#include "appframework/actormanager.h"
#include "basegamefeature/managers/sceneschedulemanager.h"
#include "app/physXfeature/physicsCore/PhysicsServer.h"

namespace App
{

	__ImplementClass(App::RootActor, 'ROAC' , App::Actor);


	RootActor::RootActor()
		:mScene(NULL)
	{
		mName.Format("__scene_name_%d",(int)this);
		SetTagID(0xffffffff);
	}

	RootActor::~RootActor()
	{

	}

	void RootActor::SetScene(Scene* scene)
	{
		mScene = scene;
	}

	Graphic::RenderScene* RootActor::GetRenderScene() const
	{
		n_assert(mScene);
		return mScene->GetRenderScene();
	}

	__ImplementClass(App::Scene, 'SCEN' ,Core::RefCounted);

	//------------------------------------------------------------------------
	Scene::Scene()
		: mModified(false)
		, mName("NoName")
		, mRenderScene(NULL)
	{
	}
	//------------------------------------------------------------------------
	Scene::~Scene()
	{
		n_assert( !mRoot.isvalid() );
		if (mRenderScene)
		{
			mRenderScene->Destroy();
			n_delete(mRenderScene);
			mRenderScene = NULL;
		}
	}
	//------------------------------------------------------------------------ 
	Graphic::RenderScene*
	Scene::GetRenderScene() const
	{
		if (mRenderScene)
		{
			return mRenderScene;
		}
		else
		{
			return SceneScheduleManager::Instance()->_GetMainRenderScene();
		}
	}
	//------------------------------------------------------------------------ 
	void
	Scene::_ApplyEnvironment(bool apply /* = false */)
	{
		GetRenderScene()->SetEnvironment(&mEnvironment);
        
#ifndef __PHYSX_COMMIT__
		if(apply)
		{
			PhysicsServer::Instance()->SetGravity(mEnvironment.graivty);
			PhysicsServer::Instance()->SetSleepThreshold(mEnvironment.sleepVel);
			PhysicsServer::Instance()->SetMaxAngularVelocity(mEnvironment.maxAngular);
			PhysicsServer::Instance()->SetDefaultMaterial(mEnvironment.defaultMat);
			if(mEnvironment.layerIDArray.Size() == 32)
			{
				PhysicsServer::Instance()->SetAllCollisionMaskMatrix(mEnvironment.layerIDArray);
			}
		}
#endif
	}

	//------------------------------------------------------------------------
	SizeT 
	Scene::GetActorCount(void) const
	{
		n_assert( mRoot.isvalid() );
		return mRoot->GetChildCount();
	}
	//------------------------------------------------------------------------ 
	const GPtr<Actor>& 
	Scene::GetActor(IndexT index) const
	{
		n_assert( mRoot.isvalid() );
		return mRoot->GetChild( index );
	}
	//------------------------------------------------------------------------
	const GPtr<Actor>& 
	Scene::FindActor( Actor::FastId actorID ) const
	{
		n_assert( mRoot.isvalid() );
		return mRoot->FindChild(actorID);
	}
	//------------------------------------------------------------------------
	const GPtr<Actor>& Scene::FindActorByTag( App::TagID id ) const
	{
		n_assert( mRoot.isvalid() );
		return mRoot->FindChildByTag(id);
	}
	//------------------------------------------------------------------------
	void Scene::FindActorsByTag(const App::TagID tagID,Util::Array< GPtr<Actor> >& actors) const
	{
		n_assert( mRoot.isvalid());
		mRoot->FindChildrenByTag(tagID,actors);
		
	}
	//------------------------------------------------------------------------
	const GPtr<Actor>& 
		Scene::FindActor( const Util::Guid& guid ) const
	{
		n_assert( mRoot.isvalid() );
		return mRoot->FindChild(guid);
	}
	//------------------------------------------------------------------------
	const GPtr<Actor>& 
		Scene::FindActor( const Util::String& name ) const
	{
		n_assert( mRoot.isvalid() );
		return mRoot->FindChild(name);
	}
	//------------------------------------------------------------------------
	IndexT 
	Scene::FindActorIndex( Actor::FastId fastID ) const
	{
		n_assert( mRoot.isvalid() );
		return mRoot->FindChildIndex(fastID);
	}
	//------------------------------------------------------------------------
	void 
	Scene::AddActor( GPtr<Actor>& pActor )
	{
		//当导入角色动画等原本就有层级结构的Actor时，就不能把他们挂在最顶层的mRoot上
		if (pActor->GetParent())
		{
			return;
		}
		n_assert( mRoot.isvalid() );
		mRoot->AddChild( pActor );
	}
	//------------------------------------------------------------------------
	void 
	Scene::RemoveActor( IndexT index)
	{
		n_assert( mRoot.isvalid() );
		return mRoot->RemoveChild(index);
	}
	//------------------------------------------------------------------------
	void 
	Scene::RemoveActor( GPtr<Actor>& pActor )
	{
		n_assert( mRoot.isvalid() );
		return mRoot->RemoveChild( pActor );
	}
	//------------------------------------------------------------------------
	void
	Scene::_Open( bool privateScene )
	{
		n_assert( !mRoot.isvalid() );

		if (privateScene)
		{
			mRenderScene = n_new(Graphic::RenderScene);
			mRenderScene->Setup();
		}

		mRoot = RootActor::Create();
		mRoot->SetScene(this);
		mRoot->Active(false);
	}
	//------------------------------------------------------------------------
	// called before close
	void 
	Scene::_Close()
	{
		if ( mRoot.isvalid() )
		{
			//mRoot->Destory(true);
			GPtr<Actor> ac = mRoot.upcast<Actor>();
			ActorManager::Instance()->DeactiveActorDelayed(ac, true);
			mRoot = NULL;
		}
	}
}
